File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic/BoundAsNewLocalDeclarations.cs
using System.Collections.Immutable;
using System.Diagnostics;

namespace Microsoft.CodeAnalysis.VisualBasic
{
	internal sealed class BoundAsNewLocalDeclarations : BoundLocalDeclarationBase
	{
		private readonly ImmutableArray<BoundLocalDeclaration> _LocalDeclarations;

		private readonly BoundExpression _Initializer;

		public ImmutableArray<BoundLocalDeclaration> LocalDeclarations => _LocalDeclarations;

		public BoundExpression Initializer => _Initializer;

		public BoundAsNewLocalDeclarations(SyntaxNode syntax, ImmutableArray<BoundLocalDeclaration> localDeclarations, BoundExpression initializer, bool hasErrors = false)
			: base(BoundKind.AsNewLocalDeclarations, syntax, hasErrors || BoundNodeExtensions.NonNullAndHasErrors(localDeclarations) || BoundNodeExtensions.NonNullAndHasErrors(initializer))
		{
			_LocalDeclarations = localDeclarations;
			_Initializer = initializer;
		}

		[DebuggerStepThrough]
		public override BoundNode Accept(BoundTreeVisitor visitor)
		{
			return visitor.VisitAsNewLocalDeclarations(this);
		}

		public BoundAsNewLocalDeclarations Update(ImmutableArray<BoundLocalDeclaration> localDeclarations, BoundExpression initializer)
		{
			if (localDeclarations != LocalDeclarations || initializer != Initializer)
			{
				BoundAsNewLocalDeclarations boundAsNewLocalDeclarations = new BoundAsNewLocalDeclarations(base.Syntax, localDeclarations, initializer, base.HasErrors);
				boundAsNewLocalDeclarations.CopyAttributes(this);
				return boundAsNewLocalDeclarations;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/BoundAsNewLocalDeclarations.vb
Imports Microsoft.CodeAnalysis
Imports System
Imports System.Collections.Immutable
Imports System.Diagnostics

Namespace Microsoft.CodeAnalysis.VisualBasic
	Friend NotInheritable Class BoundAsNewLocalDeclarations
		Inherits BoundLocalDeclarationBase
		Private ReadOnly _LocalDeclarations As ImmutableArray(Of BoundLocalDeclaration)

		Private ReadOnly _Initializer As BoundExpression

		Public ReadOnly Property Initializer As BoundExpression
			Get
				Return Me._Initializer
			End Get
		End Property

		Public ReadOnly Property LocalDeclarations As ImmutableArray(Of BoundLocalDeclaration)
			Get
				Return Me._LocalDeclarations
			End Get
		End Property

		Public Sub New(ByVal syntax As SyntaxNode, ByVal localDeclarations As ImmutableArray(Of BoundLocalDeclaration), ByVal initializer As BoundExpression, Optional ByVal hasErrors As Boolean = False)
			MyBase.New(BoundKind.AsNewLocalDeclarations, syntax, If(hasErrors OrElse localDeclarations.NonNullAndHasErrors(), True, initializer.NonNullAndHasErrors()))
			Me._LocalDeclarations = localDeclarations
			Me._Initializer = initializer
		End Sub

		<DebuggerStepThrough>
		Public Overrides Function Accept(ByVal visitor As BoundTreeVisitor) As BoundNode
			Return visitor.VisitAsNewLocalDeclarations(Me)
		End Function

		Public Function Update(ByVal localDeclarations As ImmutableArray(Of BoundLocalDeclaration), ByVal initializer As BoundExpression) As BoundAsNewLocalDeclarations
			Dim boundAsNewLocalDeclaration As BoundAsNewLocalDeclarations
			If (localDeclarations <> Me.LocalDeclarations OrElse initializer <> Me.Initializer) Then
				Dim boundAsNewLocalDeclaration1 As BoundAsNewLocalDeclarations = New BoundAsNewLocalDeclarations(MyBase.Syntax, localDeclarations, initializer, MyBase.HasErrors)
				boundAsNewLocalDeclaration1.CopyAttributes(Me)
				boundAsNewLocalDeclaration = boundAsNewLocalDeclaration1
			Else
				boundAsNewLocalDeclaration = Me
			End If
			Return boundAsNewLocalDeclaration
		End Function
	End Class
End Namespace